### ISSUE TITLE
[VMAgent] label exported_ lost randomly on timeseries

### ISSUE THREAD

**Issue by jplanckeel** [VictoriaMetrics/VictoriaMetrics#3278]
**Describe the bug**

Victoria Metrics version : 1.82.1
Kubernetes Version : v1.21.14

We have upgrade chart victoria-metrics/victoria-metrics-k8s-stack version: 0.12.6, app version 1.80.0 to version: 0.12.7, app version 1.82.1. Since we have on metrics lost this  label exported_namespace on random timeseries.

We have a metrics prescaling_metric for a timeserie A, we have label exported_namespace, for serie B we don't have label. 

![image](https://user-images.githubusercontent.com/78090694/198068117-7441d87d-c9fa-4a8a-bbdb-a56b6abac81b.png)

We have check label  is present on expoter. We have one replica of this exporter. 

See how to setup monitoring here:

Config VMagent for prescaling expoter. We use CRD VMMserviceScrape

```
- job_name: serviceScrape/prescaling-exporter/prescaling-exporter/0
  scrape_timeout: 10s
  metrics_path: /metrics
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_app_kubernetes_io_instance]
    regex: prescaling-exporter
    action: keep
  - source_labels: [__meta_kubernetes_service_label_app_kubernetes_io_name]
    regex: prescaling-exporter
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    regex: metrics
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    target_label: node
    regex: Node;(.*)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    target_label: pod
    regex: Pod;(.*)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_pod_name]
    target_label: pod
  - source_labels: [__meta_kubernetes_pod_container_name]
    target_label: container
  - source_labels: [__meta_kubernetes_namespace]
    target_label: namespace
  - source_labels: [__meta_kubernetes_service_name]
    target_label: service
  - source_labels: [__meta_kubernetes_service_name]
    target_label: job
    replacement: ${1}
  - source_labels: [__meta_kubernetes_service_label_prescaling_exporter]
    target_label: job
    regex: (.+)
    replacement: ${1}
  - target_label: endpoint
    replacement: metrics
  - source_labels: [__meta_kubernetes_pod_name]
    target_label: kubernetes_pod_name
    regex: (.*)
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    target_label: kubernetes_namespace
    regex: (.*)
    replacement: $1
    action: replace
  - regex: __meta_kubernetes_pod_label_(.+)
    action: labelmap
  - regex: (endpoint|pod_template_hash|pod_template_generation|controller_revision_hash)
    action: labeldrop
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    namespaces:
      own_namespace: false
      names:
      - prescaling-exporter
```

**Comment by jplanckeel** [comment]
We have rollback to app version 1.80.0 this solve the problem.

**Comment by jplanckeel** [comment]
i have question on functions appendLabel and appendExtraLabel, i have create a [PR](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3281) to exchange on this

**Comment by valyala** [comment]
@jplanckeel , thanks for the bugreport and the initial attempt to fix it! The issue was in the `labels := dst[offset:]` line at the top of `appendExtraLabels()` function. The `labels` slice contains the initial metric labels, where `extraLabels` must be added. When `extraLabels` are added to `dst`, it may be re-allocated if it has no enough capacity. In this case the `labels` start pointing to the previous array, which was stored at `dst` before the re-allocation. So, all the changes inside the `labels` slice (such as addition of `exported_` prefix) become invisible to the caller of `appendExtraLables()` function. The fix is to create `labels` slice at the beginning of the loop over `extraLabels`. See [this commit](76e8888272589a65d93a5d1e6417ab4c954ee5c4) for details.

**Comment by valyala** [comment]
The bugfix mentioned [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3278#issuecomment-1295368078) has been included in [vmagent v1.83.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue as fixed then.

**Comment by jplanckeel** [comment]
@valyala  Thank you for you reponse :)
